common.go

```diff
@@ -0,0 +1,110 @@
+package cose
+
+import (
+	"errors"
+	"fmt"
+)
+
+// IntOrStr is a value that can be
+type IntOrStr struct {
+	intVal *int
```

Comment:
should this be `int64`?

Reply:
On top of what @thomas-fossati said, we could avoid allocating when the wrapped value is an integer by not using a pointer on `intVal`. Calling `strVal != nil` is enough to know if the value being wrapped is an integer or a string.

Reply:
Even better, just don't use pointers at all and add a discriminator property, i.e. `isstring`. This way we avoid allocating if the wrapped value is a string.

Reply:
Sorry, could you clarify  what you mean?  If the wrapped value is a string, there would still be an allocation during the unmarshalling. I'm not sure what we'd be saving here?

Reply:
Sure! I'm thinking in something like this:

```go

type IntOrStr struct {
    intVal int64
    strVal string
    isString bool
}

func (ios *IntOrStr) Set(v interface{}) error {
	switch t := v.(type) {
	case int:
		ios.intVal = t
		ios.strVal = nil
		ios.isString= false
	case string:
		ios.strVal = t
		ios.intVal = nil
		ios.isString= true
	default:
		return fmt.Errorf("must be int or string, found %T", t)
	}

	return nil
}

func (ios IntOrStr) IsString() bool {
	returnios.isString
}

func (ios IntOrStr) Int() int64 {
	if !ios.isString {
		return ios.intVal
	}
	return 0
}
...
```

Reply:
Thanks! Could you explain why this would be preferable? This seem to just add an extra field to the struct, but on not sure how this prevents an allocation that would take place otherwise.

Reply:
Ah, gotcha. Thanks for the detail expalanantion! I'll make the change.